File: test/vectors/cross-emitter-matrix/verify.mjs
#!/usr/bin/env node
// Cross-emitter matrix — TypeScript/Node.js verifier.
// Reads {record, signature_hex, public_key_hex} JSON from stdin.
// Locally recomputes canonical form from record, verifies signature against it.
// Writes {verified: bool, local_canonical, sig_hex} to stdout.

import { verify as nodeVerify, createPublicKey } from "node:crypto";

const RECORD_FIELD_ORDER = [
  "id", "timestamp", "method", "toolName", "namespace", "upstream",
  "principal", "durationMs", "success", "errorCode", "previousHash",
];

function assertWellFormed(value) {
  for (let i = 0; i < value.length; i++) {
    const code = value.charCodeAt(i);
    if (code >= 0xD800 && code <= 0xDBFF) {
      const next = i + 1 < value.length ? value.charCodeAt(i + 1) : 0;
      if (next < 0xDC00 || next > 0xDFFF) throw new Error(`unpaired surrogate at ${i}`);
      i++;
    } else if (code >= 0xDC00 && code <= 0xDFFF) {
      throw new Error(`unpaired surrogate at ${i}`);
    }
  }
}

function canonicalizeValue(value) {
  if (value === null || value === undefined) return null;
  if (typeof value === "string") { assertWellFormed(value); return value; }
  if (typeof value === "boolean") return value;
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) throw new Error(`unsafe number ${value}`);
    return value;
  }
  if (Array.isArray(value)) return ["L", value.map(canonicalizeValue)];
  if (typeof value === "object") {
    const keys = Object.keys(value).sort().filter((k) => value[k] !== undefined);
    return ["M", keys.map((k) => [k, canonicalizeValue(value[k])])];
  }
  throw new Error(`unsupported type ${typeof value}`);
}

function canonicalize(record) {
  const ordered = RECORD_FIELD_ORDER.map((k) => [k, canonicalizeValue(record[k] ?? null)]);
  for (const optional of ["decisionContextDigest", "extensionsDigest", "aiInvocation", "parties"]) {
    if (record[optional] != null) {
      ordered.push([optional, canonicalizeValue(record[optional])]);
    }
  }
  return JSON.stringify(ordered);
}

function publicKeyFromHex(publicKeyCache, publicKeyHex) {
  const cached = publicKeyCache.get(publicKeyHex);
  if (cached) return cached;
  const pubDer = Buffer.concat([
    Buffer.from("302a300506032b6570032100", "hex"),
    Buffer.from(publicKeyHex, "hex"),
  ]);
  const publicKey = createPublicKey({ key: pubDer, format: "der", type: "spki" });
  publicKeyCache.set(publicKeyHex, publicKey);
  return publicKey;
}

async function runOneShot() {
  const chunks = [];
  for await (const chunk of process.stdin) chunks.push(chunk);
  const { record, signature_hex, public_key_hex } = JSON.parse(Buffer.concat(chunks).toString("utf-8"));

  const pubDer = Buffer.concat([
    Buffer.from("302a300506032b6570032100", "hex"),
    Buffer.from(public_key_hex, "hex"),
  ]);
  const publicKey = createPublicKey({ key: pubDer, format: "der", type: "spki" });

  let localCanonical, verified;
  try {
    localCanonical = canonicalize(record);
    verified = nodeVerify(null, Buffer.from(localCanonical, "utf-8"), publicKey, Buffer.from(signature_hex, "hex"));
  } catch (e) {
    verified = false;
    localCanonical = `ERROR: ${e.message}`;
  }
  process.stdout.write(JSON.stringify({ verified, local_canonical: localCanonical, sig_hex: signature_hex }));
}

function writeLineFlushed(line) {
  return new Promise((resolve, reject) => {
    process.stdout.write(line + "\n", (err) => (err ? reject(err) : resolve()));
  });
}

async function runDaemon() {
  // Env pinning — only set if missing so we do not override an explicit choice.
  if (!process.env.LC_ALL) process.env.LC_ALL = "C.UTF-8";
  if (!process.env.PYTHONIOENCODING) process.env.PYTHONIOENCODING = "utf-8";

  // Cache public keys across records — every driver run reuses the same key,
  // and this avoids a fresh SPKI parse per line.
  const publicKeyCache = new Map();

  const { createInterface } = await import("node:readline");
  const rl = createInterface({ input: process.stdin, crlfDelay: Infinity });

  for await (const line of rl) {
    if (!line) continue;
    let id = null;
    try {
      const req = JSON.parse(line);
      id = req.id ?? null;
      const { record, signature_hex, public_key_hex } = req;
      const publicKey = publicKeyFromHex(publicKeyCache, public_key_hex);
      const localCanonical = canonicalize(record);
      const verified = nodeVerify(
        null,
        Buffer.from(localCanonical, "utf-8"),
        publicKey,
        Buffer.from(signature_hex, "hex"),
      );
      await writeLineFlushed(JSON.stringify({ id, ok: true, verified, local_canonical: localCanonical }));
    } catch (e) {
      try {
        await writeLineFlushed(JSON.stringify({ id, ok: false, error: String(e && e.message ? e.message : e) }));
      } catch (_) {
        // stdout gone; nothing to do — loop terminates on next iteration EOF.
      }
    }
  }
}

async function main() {
  if (process.env.DAEMON_MODE === "1") {
    await runDaemon();
  } else {
    await runOneShot();
  }
}
main().catch((e) => { console.error(e.message); process.exit(1); });

File: test/vectors/cross-emitter-matrix/verify.py
#!/usr/bin/env python3
# Cross-emitter matrix — Python verifier.
# Reads {record, signature_hex, public_key_hex} JSON from stdin.
# Locally recomputes canonical form from record, verifies signature against it.
# Writes {verified: bool, local_canonical, sig_hex} to stdout.

import json
import os
import sys
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
from cryptography.exceptions import InvalidSignature


FIELD_ORDER = [
    "id", "timestamp", "method", "toolName", "namespace", "upstream",
    "principal", "durationMs", "success", "errorCode", "previousHash",
]


def assert_well_formed(value):
    for i, ch in enumerate(value):
        if 0xD800 <= ord(ch) <= 0xDFFF:
            raise ValueError(f"unpaired surrogate at index {i}")


def canonicalize_value(value):
    if value is None:
        return None
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        assert_well_formed(value)
        return value
    if isinstance(value, int):
        if abs(value) > 2**53 - 1:
            raise ValueError(f"unsafe number {value}")
        return value
    if isinstance(value, float):
        raise ValueError(f"unsafe number {value}")
    if isinstance(value, list):
        return ["L", [canonicalize_value(v) for v in value]]
    if isinstance(value, dict):
        keys = sorted(value.keys(), key=lambda k: k.encode("utf-16-be"))
        for k in keys:
            assert_well_formed(k)
        return ["M", [[k, canonicalize_value(value[k])] for k in keys]]
    raise ValueError(f"unsupported type {type(value)}")


def canonicalize(record):
    ordered = [[k, canonicalize_value(record.get(k))] for k in FIELD_ORDER]
    for optional in ("decisionContextDigest", "extensionsDigest", "aiInvocation", "parties"):
        if record.get(optional) is not None:
            ordered.append([optional, canonicalize_value(record[optional])])
    return json.dumps(ordered, ensure_ascii=False, separators=(",", ":"))


def daemon_main():
    # Env pinning: only set fallbacks; do not override the driver.
    os.environ.setdefault("LC_ALL", "C.UTF-8")
    os.environ.setdefault("PYTHONIOENCODING", "utf-8")

    # Cache Ed25519PublicKey objects keyed by hex — the whole point of the
    # daemon is to avoid rebuilding crypto state on every record.
    pk_cache = {}

    def _get_pk(pk_hex):
        pk = pk_cache.get(pk_hex)
        if pk is None:
            pk = Ed25519PublicKey.from_public_bytes(bytes.fromhex(pk_hex))
            pk_cache[pk_hex] = pk
        return pk

    while True:
        line = sys.stdin.readline()
        if not line:
            break  # EOF — exit loop, terminate daemon
        line = line.strip()
        if not line:
            continue  # skip blank keep-alive lines
        req_id = None
        try:
            req = json.loads(line)
            req_id = req.get("id")
            record = req["record"]
            sig_hex = req["signature_hex"]
            pk_hex = req["public_key_hex"]
            public_key = _get_pk(pk_hex)

            try:
                local_canonical = canonicalize(record)
                public_key.verify(
                    bytes.fromhex(sig_hex), local_canonical.encode("utf-8")
                )
                verified = True
            except InvalidSignature:
                verified = False
            except Exception as e:
                verified = False
                local_canonical = f"ERROR: {e}"

            resp = {
                "id": req_id,
                "ok": True,
                "verified": verified,
                "local_canonical": local_canonical,
                "sig_hex": sig_hex,
            }
        except Exception as e:
            # Request-level failure (bad JSON, missing field, malformed hex).
            resp = {"id": req_id, "ok": False, "error": str(e)}
        sys.stdout.write(json.dumps(resp) + "\n")
        sys.stdout.flush()


def main():
    if os.environ.get("DAEMON_MODE") == "1":
        daemon_main()
        return

    # One-shot mode: byte-identical to pre-patch behavior.
    payload = json.loads(sys.stdin.read())
    record = payload["record"]
    sig_hex = payload["signature_hex"]
    pk_hex = payload["public_key_hex"]

    public_key = Ed25519PublicKey.from_public_bytes(bytes.fromhex(pk_hex))

    try:
        local_canonical = canonicalize(record)
        public_key.verify(bytes.fromhex(sig_hex), local_canonical.encode("utf-8"))
        verified = True
    except InvalidSignature:
        verified = False
    except Exception as e:
        verified = False
        local_canonical = f"ERROR: {e}"

    print(json.dumps({"verified": verified, "local_canonical": local_canonical, "sig_hex": sig_hex}))


if __name__ == "__main__":
    main()
